### PR TITLE
GH-125789: fix `fut._callbacks` to always return a copy of callbacks

### PR DESCRIPTION
**Pull request by kumaraditya303**
* Issue: gh-125789

### CODE REVIEW

File: Misc/NEWS.d/next/Library/2024-10-24-14-08-10.gh-issue-125789.eaiAMw.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Fix possible crash when mutating list of callbacks returned by :attr:`!asyncio.Future._callbacks`. It now always returns a new copy in C implementation :mod`!_asyncio`. Patch by Kumar Aditya.
```
Comment (kumaraditya303) [inline Misc/NEWS.d/next/Library/2024-10-24-14-08-10.gh-issue-125789.eaiAMw.rst]:
```suggestion
Fix possible crash when mutating list of callbacks returned by :attr:`!asyncio.Future._callbacks`. It now always returns a new copy in C implementation :mod:`!_asyncio`. Patch by Kumar Aditya.
```

File: Modules/_asynciomodule.c [hunk 2]
```diff
@@ -1265,52 +1265,49 @@ static PyObject *
...
-    if (fut->fut_callback0 == NULL) {
-        if (fut->fut_callbacks == NULL) {
-            Py_RETURN_NONE;
-        }
-
-        return Py_NewRef(fut->fut_callbacks);
+    Py_ssize_t len = 0;
+    if (fut->fut_callback0 != NULL) {
+        len++;
     }
```
Comment (picnixz) [inline Modules/_asynciomodule.c:1273]:
```suggestion
    Py_ssize_t len = fut->fut_callback0 == NULL ? 0 : 1;
```

File: Modules/_asynciomodule.c [hunk 3]
```diff
@@ -1265,52 +1265,49 @@ static PyObject *
...
+        assert(fut->fut_context0 != NULL);
+        PyTuple_SET_ITEM(tup0, 1, Py_NewRef(fut->fut_context0));
+        PyList_SET_ITEM(callbacks, i, tup0);
+        i++;
+    }
 
     if (fut->fut_callbacks != NULL) {
-        for (i = 0; i < PyList_GET_SIZE(fut->fut_callbacks); i++) {
-            PyObject *cb = PyList_GET_ITEM(fut->fut_callbacks, i);
+        for (int j = 0; j < PyList_GET_SIZE(fut->fut_callbacks); j++) {
```
Comment (picnixz) [inline Modules/_asynciomodule.c]:
```suggestion
        for (Py_ssize_t j = 0; j < PyList_GET_SIZE(fut->fut_callbacks); j++) {
```

File: Lib/test/test_asyncio/test_futures.py [hunk 4]
```diff
@@ -697,6 +697,22 @@ def test_future_del_segfault(self):
...
+        fut = self._new_future(loop=self.loop)
+        f1 = lambda _: 1
+        f2 = lambda _: 2
+        fut.add_done_callback(f1)
+        fut.add_done_callback(f2)
+        callbacks = fut._callbacks
+        self.assertIsNot(callbacks, fut._callbacks)
+        fut.remove_done_callback(f1)
+        callbacks = fut._callbacks
+        self.assertIsNot(callbacks, fut._callbacks)
```
Comment (picnixz) [inline Lib/test/test_asyncio/test_futures.py:714]: How about removing f2 as well and check that the returned list is None in this case?

**Review by picnixz — COMMENTED** [review]
I would add a test for that (e.g., one of the test that Nico wrote with the evil classes) to see that the interpreter does not crash anymore.

EDIT: My previous comment was wrong. We can still change the length by just adding new callbacks but we cannot change the type of the items.

**Comment by kumaraditya303** [comment]
> I would add a test for that (e.g., one of the test that Nico wrote with the evil classes) to see that the interpreter does not crash anymore.

I think it is more appropriate  to add that to your PR which fixes the UAF issues.

File: Modules/_asynciomodule.c (hunk 2 above)
Comment (kumaraditya303) [reply Modules/_asynciomodule.c:1273]: doesn't seem much better, and if is more readable to I'm sticking to it

**Pull request merged**